platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/ProfileMigratable.java

```diff
@@ -0,0 +1,280 @@
+   * might have side effects and install or start others which were not supposed to in the original
+   * system. As such, we are going to attempt multiple times for each operations such that we can
+   * account for side effects of other operations. In addition, because we are unable to determine a
+   * proper order to process all features or bundles, it is possible that we end up attempting to
+   * install a feature for which the dependencies have not yet been installed. We are therefore
+   * forced to attempt multiple times hoping that dependencies are properly being handled by the
+   * previous pass. Failures will be reported only on the last attempt. Retries will automatically
+   * stop when all apps, features, and apps are determined to be as they were at export time.
+   */
+  @SuppressWarnings("PMD.DefaultPackage" /* designed to be used by TaskList within this package */)
+  static final int RETRY_COUNT = 5;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ProfileMigratable.class);
+
+  /**
+   * Holds the current export version.
+   *
+   * <p>1.0 - initial version
+   */
+  private static final String VERSION = "1.0";
+
+  private static final Path PROFILE_PATH = Paths.get("profile.json");
+
+  @VisibleForTesting static final ObjectMapper MAPPER = JsonFactory.create();
+
+  private final ApplicationManager appMgr;
+
+  private final FeatureManager featureMgr;
+
+  private final BundleManager bundleMgr;
+
+  @VisibleForTesting
+  ProfileMigratable(ApplicationManager appMgr, FeatureManager featureMgr, BundleManager bundleMgr) {
+    this.appMgr = appMgr;
+    this.featureMgr = featureMgr;
+    this.bundleMgr = bundleMgr;
+  }
+
+  /**
+   * Instantiates a profile migratable with the specified application and features services.
+   *
+   * @param appService the application services to use
+   * @param featuresService the features service to use
+   * @throws IllegalArgumentException if <code>applicationService</code> or <code>featuresService
+   * </code> is <code>null</code>
+   */
+  public ProfileMigratable(ApplicationService appService, FeaturesService featuresService) {
+    this(
+        new ApplicationManager(appService),
+        new FeatureManager(featuresService),
+        new BundleManager());
+  }
+
+  @Override
+  public String getVersion() {
+    return ProfileMigratable.VERSION;
+  }
+
+  @Override
+  public String getId() {
+    return "ddf.profile";
+  }
+
+  @Override
+  public String getTitle() {
+    return "Profile Migratable";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Exports the state of applications, features, and bundles";
+  }
+
+  @Override
+  public String getOrganization() {
+    return "Codice";
+  }
+
+  @Override
+  public void doExport(ExportMigrationContext context) {
+    context
+        .getEntry(ProfileMigratable.PROFILE_PATH)
+        .store(
+            (r, os) ->
+                ProfileMigratable.MAPPER.writeValue(
```

Comment:
❓ Instead of having to deal with maps of maps and lists and values, could we have created a top-level bean (e.g., `JsonInstallationProfile`), re-used the `JsonApplication`, etc., as fields inside it, and use those to serialize/deserialize the data? Boon's JSON Mapper has a nice feature that allows serialization and deserialization of Java objects instead of using raw types like maps, lists, etc.

Reply:
I didn't think Boon supported that. I used this approach based on what @brjeter had done in the profile install command originally. I definitely prefer a more OO approach which avoids doing all this map manipulation and address a lot of the previous comments for the JsonUtils class.